container/kvm/cache_internal_test.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+package kvm
+
+import (
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/juju/juju/environs/imagedownloads"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+)
+
+// internalCacheSuite is gocheck boilerplate.
+type internalCacheSuite struct{}
+
+var _ = gc.Suite(internalCacheSuite{})
+
+const imageContents = "fake img file"
+
+func (internalCacheSuite) TestUpdater(c *gc.C) {
+	mtime := time.Unix(1000, 0).UTC()
+	imageFile := &fakeFileInfo{
+		basename: "series-image.img",
+		modtime:  mtime,
+		contents: imageContents,
+	}
+	fs := fakeFS{
+		"/": &fakeFileInfo{
+			dir:  true,
+			ents: []*fakeFileInfo{imageFile},
+		},
+		"/server.img": imageFile,
+	}
+	ts := httptest.NewServer(http.FileServer(fs))
+	defer ts.Close()
+
+	md := newTestMetadata(ts.URL)
+
+	tmpdir, pathfinder, ok := newTmpdir()
+	if !ok {
+		c.Fatal("failed to setup temp dir in test")
+	}
+	defer func() {
+		err := os.RemoveAll(tmpdir)
+		if err != nil {
+			c.Errorf("got error %q when removing tmpdir %q",
+				err.Error(),
+				tmpdir)
+		}
+	}()
+
+	updater, err := newDefaultUpdater(md, pathfinder)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Test 200 returns reader, nil
+	err = updater.Update()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Test 304 returns nil,nil and that mtimes don't change.
+	finfo, err := os.Stat(tmpdir + "/kvm/images/" + md.Path)
+	if err != nil {
+		c.Fatalf("failed to stat expected file: %s", err.Error())
+	}
+	pinfo, err := os.Stat(tmpdir + "/kvm/metadata/" + md.Path)
+	if err != nil {
+		c.Fatalf("failed to stat expected file: %s", err.Error())
+	}
+
+	updater, err = newDefaultUpdater(md, pathfinder)
+	err = updater.Update()
+	c.Assert(err, jc.ErrorIsNil)
+
+	finfo2, err := os.Stat(tmpdir + "/kvm/images/" + md.Path)
+	if err != nil {
+		c.Fatalf("failed to stat expected file: %s", err.Error())
+	}
+	pinfo2, err := os.Stat(tmpdir + "/kvm/metadata/" + md.Path)
+	if err != nil {
+		c.Fatalf("failed to stat expected file: %s", err.Error())
+	}
+
+	c.Check(finfo.ModTime(), jc.DeepEquals, finfo2.ModTime())
+	c.Check(pinfo.ModTime(), jc.DeepEquals, pinfo2.ModTime())
+
+	// Test setting the time forward on the upstream image to test updating an
+	// existing image. Also ensure the mtime on the files changes.
+	time.Sleep(5 * time.Millisecond) // need enough time to flush the changes to disk.
+
+	imageFile.modtime = imageFile.modtime.Add(1 * time.Hour)
+	updater, err = newDefaultUpdater(md, pathfinder)
+	updater.fileCache.ModTime = mtime.Format(http.TimeFormat)
+	err = updater.Update()
+	c.Assert(err, jc.ErrorIsNil)
+
+	finfo2, err = os.Stat(tmpdir + "/kvm/images/" + md.Path)
```

Comment:
I do not think that these tests will pass on windows.. file separators sensitivities :)